File: src/java/2_CreationalPatterns/3_Prototype/CopyThroughSerialization.java
package com.activemesa.creational.prototype.copyserial;

import org.apache.commons.lang3.SerializationUtils;

import java.io.Serializable;

// some libraries use reflection (no need for Serializable)
class Foo implements Serializable
{
  public int stuff;
  public String whatever;

  public Foo(int stuff, String whatever)
  {
    this.stuff = stuff;
    this.whatever = whatever;
  }

  @Override
  public String toString()
  {
    return "Foo{" +
      "stuff=" + stuff +
      ", whatever='" + whatever + '\'' +
      '}';
  }
}

class CopyThroughSerializationDemo
{
  public static void main(String[] args)
  {
    Foo foo = new Foo(42, "life");
    // use apache commons!
    Foo foo2 = SerializationUtils.roundtrip(foo);

    foo2.whatever = "xyz";

    System.out.println(foo);
    System.out.println(foo2);
  }
}

File: src/csharp/2_CreationalPatterns/3_Prototype/CopyThroughSerialization.cs
using System;
using static System.Console;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using System.Xml.Serialization;

namespace DotNetDesignPatternDemos.Creational.Prototype
{
  public static class ExtensionMethods
  {
    public static T DeepCopy<T>(this T self)
    {
      MemoryStream stream = new MemoryStream();
      BinaryFormatter formatter = new BinaryFormatter();
      formatter.Serialize(stream, self);
      stream.Seek(0, SeekOrigin.Begin);
      object copy = formatter.Deserialize(stream);
      stream.Close();
      return (T)copy;
    }

    public static T DeepCopyXml<T>(this T self)
    {
      using (var ms = new MemoryStream())
      {
        XmlSerializer s = new XmlSerializer(typeof(T));
        s.Serialize(ms, self);
        ms.Position = 0;
        return (T) s.Deserialize(ms);
      }
    }
  }

  //[Serializable] // this is, unfortunately, required
  public class Foo
  {
    public uint Stuff;
    public string Whatever;

    public override string ToString()
    {
      return $"{nameof(Stuff)}: {Stuff}, {nameof(Whatever)}: {Whatever}";
    }
  }

  public static class CopyThroughSerialization
  {
    static void Main()
    {
      Foo foo = new Foo {Stuff = 42, Whatever = "abc"};

      //Foo foo2 = foo.DeepCopy(); // crashes without [Serializable]
      Foo foo2 = foo.DeepCopyXml();

      foo2.Whatever = "xyz";
      WriteLine(foo);
      WriteLine(foo2);
    }
  }
}
